File: AndroidStudioProjects/app/src/main/java/cn/zemic/hy/display/unmannedstoragedisplay/presenter/imp/alarm/AlarmRepository.java
package cn.zemic.hy.display.unmannedstoragedisplay.presenter.imp.alarm;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutionException;

import cn.zemic.hy.display.unmannedstoragedisplay.model.bean.FilterLedForMobileUiCommand;
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.EntranceGuardState;
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.MachineBindDoorControlViewModel;
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.OperateWarningInformViewModel;
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.ShelfState;
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.UserInViewModel;
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.UserOutVM;
import cn.zemic.hy.display.unmannedstoragedisplay.network.URLFactory;
import cn.zemic.hy.display.unmannedstoragedisplay.network.service.FetchDataService;
import cn.zemic.hy.display.unmannedstoragedisplay.presenter.interfaces.IAlarmRepository;
import cn.zemic.hy.display.unmannedstoragedisplay.presenter.interfaces.IBaseRepository;
import microsoft.aspnet.signalr.client.Platform;
import microsoft.aspnet.signalr.client.SignalRFuture;
import microsoft.aspnet.signalr.client.http.android.AndroidPlatformComponent;
import microsoft.aspnet.signalr.client.hubs.HubConnection;
import microsoft.aspnet.signalr.client.hubs.HubProxy;
import microsoft.aspnet.signalr.client.transport.ServerSentEventsTransport;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * @author fxs
 */
public class AlarmRepository implements IAlarmRepository {
    private static final int DATA_ERROR = 0;
    private static final int DATA_CHANGED = 1;
    private static final int USER_CHANGED = 2;
    private static final int DATA_SEND_MSG = 3;
    private static final int USER_OUT = 4;
    private static final int CHECK = 5;
    private static final int MAINTAIN = 6;
    private static final int TEMPERATURE_HUMIDITY = 7;
    private static final int ENTRANCE_GUARD_STATE = 8;
    private static final int SHELF_STATE = 9;

    private static IBaseRepository.OnGetWarningDataFinish mOnGetWarningDataFinish;
    private static String mConnectionId;
    private MyMessagesHandler mHandler;
    private HubProxy mHubProxy;

    @Override
    public void getWareHouseNo(String deviceId, OnGetWareHouseNoFinish onGetWareHouseNoFinish) {
        FilterLedForMobileUiCommand filter = new FilterLedForMobileUiCommand(deviceId);
        FetchDataService.getWareHouseNoService(filter).enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {
                if (!response.isSuccessful()) {
                    onGetWareHouseNoFinish.onFail(String.format(Locale.CHINA, "获取仓库号响应异常，错误码：%d", response.code()));
                    return;
                }
                String wareHouseNo = response.body();
                onGetWareHouseNoFinish.onSuccess(wareHouseNo);
            }

            @Override
            public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {
                onGetWareHouseNoFinish.onFail(t.getMessage() == null || "".equals(t.getMessage()) ? "获取仓库号失败" : t.getMessage());
            }
        });
    }

    @Override
    public void connectServiceForGetWarningAndUserInfo(String deviceId, String wareHouseNo, IBaseRepository.OnGetWarningDataFinish onGetWarningDataFinish) {
        mOnGetWarningDataFinish = onGetWarningDataFinish;
        mHandler = new MyMessagesHandler();
        Platform.loadPlatformComponent(new AndroidPlatformComponent());
        String serverURI = String.format(Locale.CHINA, "%ssignalr", URLFactory.serverURI);
//        String serverURI = "http://192.168.0.2:25535/signalr";
        HubConnection connection = new HubConnection(serverURI) {
            @Override
            protected void onConnected() {
                super.onConnected();
            }

            @Override
            public void onError(Throwable error, boolean mustCleanCurrentConnection) {
                super.onError(error, mustCleanCurrentConnection);
                Message message = Message.obtain();
                message.what = DATA_ERROR;
                message.obj = "".equals(error.getMessage()) ? "连接推送服务器出现异常，请联系管理员" : error.getMessage();
                mHandler.sendMessage(message);
            }

            @Override
            protected void onClosed() {
                super.onClosed();
            }
        };
        //hub agency
        mHubProxy = connection.createHubProxy("electronicSignsHub");
        //hubProxy.on可以理解为讯息or事件监听器
        mHubProxy.on("sendUserIn", this::runWithUserInfo, String.class, Boolean.class, UserInViewModel.class, MachineBindDoorControlViewModel.class);
        mHubProxy.on("sendCheck", this::runWithIsCheck, String.class, Boolean.class);
        mHubProxy.on("sendMaintain", this::runWithMaintain, String.class, Boolean.class);
        mHubProxy.on("sendTemAndHum", this::runWithTemperature, String.class, Float.class, Float.class);
        mHubProxy.on("sendUserLeave", this::runWithUserOut, String.class, UserOutVM.class);
        //接收告警信息
        mHubProxy.on("sendWarnInform", this::runWithWarnInfo, String.class, OperateWarningInformViewModel.class);
        //接受日志信息
        mHubProxy.on("sendConnectMessage", this::runWithConnectMsg, String.class, String.class, Boolean.class);
        mHubProxy.on("sendAttendanceMachine", this::entranceGuardState, String.class, EntranceGuardState.class);
        mHubProxy.on("sendLedShelfAndunit", this::shelfState, String.class, ShelfState.class);
        final SignalRFuture<Void> con = connection.start(new ServerSentEventsTransport(connection.getLogger()));
        try {
            con.get();
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        con.onError(this::onError);
        if (con.isDone()) {
            //调用服务器连接方法
            mHubProxy.invoke("Connect", wareHouseNo, deviceId);
        }
    }

    @Override
    public HubProxy getHubProxy() {
        if (null == mHubProxy) {
            throw new NullPointerException("mHubProxy is null");
        }
        return mHubProxy;
    }

    private void runWithUserInfo(String warehouseNo, Boolean isOpenDoor, UserInViewModel userIn, MachineBindDoorControlViewModel doorControl) {
        Message message = Message.obtain();
        message.what = USER_CHANGED;
        Bundle bundle = new Bundle();
        bundle.putBoolean("isOpenDoor", isOpenDoor);
        bundle.putString("warehouseNo", warehouseNo);
        bundle.putParcelable("userIn", userIn);
        bundle.putParcelable("doorControl", doorControl);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }


    private void runWithUserOut(String warehouseNo, UserOutVM user) {
        Message message = Message.obtain();
        message.what = USER_OUT;
        Bundle bundle = new Bundle();
        bundle.putString("warehouseNo", warehouseNo);
        bundle.putParcelable("userOut", user);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void runWithIsCheck(String warehouseNo, Boolean isCheck) {
        Message message = Message.obtain();
        message.what = CHECK;
        Bundle bundle = new Bundle();
        bundle.putBoolean("isCheck", isCheck);
        bundle.putString("warehouseNo", warehouseNo);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void runWithMaintain(String warehouseNo, Boolean isMaintain) {
        Message message = Message.obtain();
        message.what = MAINTAIN;
        Bundle bundle = new Bundle();
        bundle.putBoolean("isMaintain", isMaintain);
        bundle.putString("warehouseNo", warehouseNo);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void runWithTemperature(String wareHouseNo, float temperature, float humidity) {
        Message message = Message.obtain();
        message.what = TEMPERATURE_HUMIDITY;
        Bundle bundle = new Bundle();
        bundle.putString("wareHouseNo", wareHouseNo);
        bundle.putFloat("temperature", temperature);
        bundle.putFloat("humidity", humidity);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void runWithWarnInfo(String warehouseNo, OperateWarningInformViewModel warningInform) {
        Message message = Message.obtain();
        message.what = DATA_CHANGED;
        Bundle bundle = new Bundle();
        bundle.putString("warehouseNo", warehouseNo);
        bundle.putParcelable("warningInform", warningInform);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void runWithConnectMsg(String connectionMsg, String connectionId, Boolean isConnect) {
        Message message = Message.obtain();
        message.what = DATA_SEND_MSG;
        Bundle bundle = new Bundle();
        if (isConnect) {
            mConnectionId = connectionId;
        }
        bundle.putString("connectionId", connectionId);
        bundle.putBoolean("isConnect", isConnect);
        message.setData(bundle);
        message.obj = connectionMsg;
        mHandler.sendMessage(message);
    }

    private void entranceGuardState(String warehouseNo, EntranceGuardState entranceGuardState) {
        Message message = Message.obtain();
        message.what = ENTRANCE_GUARD_STATE;
        Bundle bundle = new Bundle();
        bundle.putString("warehouseNo", warehouseNo);
        bundle.putParcelable("entranceGuardState", entranceGuardState);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void shelfState(String warehouseNo, ShelfState shelfState) {
        Message message = Message.obtain();
        message.what = SHELF_STATE;
        Bundle bundle = new Bundle();
        bundle.putString("warehouseNo", warehouseNo);
        bundle.putParcelable("shelfState", shelfState);
        message.setData(bundle);
        mHandler.sendMessage(message);
    }

    private void onError(Throwable error) {
        Message message = Message.obtain();
        message.what = DATA_ERROR;
        message.obj = "".equals(error.getMessage()) ? error.getMessage() : "连接中断，请联系管理员";
        mHandler.sendMessage(message);
    }

    private static class MyMessagesHandler extends Handler {
        private final int LIST_MAX_SIZE = 10;
        private List<OperateWarningInformViewModel> mOperateWarningInformViewModels = new ArrayList<>();

        @Override
        public void handleMessage(Message msg) {
            if (mOnGetWarningDataFinish == null) {
                return;
            }
            super.handleMessage(msg);
            switch (msg.what) {
                case TEMPERATURE_HUMIDITY:
                    Bundle bundletempAndHum = msg.getData();
                    float temperature = bundletempAndHum.getFloat("temperature");
                    float humidity = bundletempAndHum.getFloat("humidity");
                    String wareHouseForTemp = bundletempAndHum.getString("wareHouseNo");
                    mOnGetWarningDataFinish.onFetchTemperatureAndHumidityFinish(wareHouseForTemp, temperature, humidity);
                    break;
                case MAINTAIN:
                    Bundle bundleMaintain = msg.getData();
                    String warehouseNoForMaintain = bundleMaintain.getString("warehouseNo");
                    Boolean isMaintain = bundleMaintain.getBoolean("isMaintain");
                    mOnGetWarningDataFinish.onRunningMaintain(warehouseNoForMaintain, isMaintain);
                    break;
                case CHECK:
                    Bundle bundleCheck = msg.getData();
                    String warehouseNo = bundleCheck.getString("warehouseNo");
                    Boolean isCheck = bundleCheck.getBoolean("isCheck");
                    mOnGetWarningDataFinish.onRunningCheck(warehouseNo, isCheck);
                    break;

                case DATA_SEND_MSG:
                    Bundle bundleMsg = msg.getData();
                    String connectionId = bundleMsg.getString("connectionId");
                    Boolean isConnect = bundleMsg.getBoolean("isConnect");
                    Log.e("DATA_SEND_MSG", msg.obj.toString() + connectionId);
                    if (isConnect) {
                        mOnGetWarningDataFinish.onShow(msg.obj.toString());
                    } else {
                        if (mConnectionId.equals(connectionId)) {
                            mOnGetWarningDataFinish.onShow(msg.obj.toString());
                        }
                    }
                    break;
                case DATA_ERROR:
                    Log.e("DATA_ERROR", msg.obj.toString());
                    mOnGetWarningDataFinish.onShow(msg.obj.toString());
                    break;
                case DATA_CHANGED:
                    Bundle bundle = msg.getData();
                    OperateWarningInformViewModel wainingInfo = bundle.getParcelable("warningInform");
                    if (wainingInfo != null) {
                        if (mOperateWarningInformViewModels.size() < LIST_MAX_SIZE) {
                            mOperateWarningInformViewModels.add(0, wainingInfo);
                        } else {
                            mOperateWarningInformViewModels.remove(mOperateWarningInformViewModels.size() - 1);
                            mOperateWarningInformViewModels.add(0, wainingInfo);
                        }
                    }
                    //update view
                    if (mOperateWarningInformViewModels.size() <= 1) {
                        mOnGetWarningDataFinish.onSuccess(mOperateWarningInformViewModels, true);
                    } else {
                        mOnGetWarningDataFinish.onSuccess(mOperateWarningInformViewModels, false);
                    }
                    break;
                case USER_CHANGED:
//                    //清空告警列表
//                    mOperateWarningInformViewModels.clear();
//                    mOnGetWarningDataFinish.onSuccess(mOperateWarningInformViewModels, false);
                    Bundle userBundle = msg.getData();
                    boolean isOpenDoor = userBundle.getBoolean("isOpenDoor");
                    UserInViewModel userIn = userBundle.getParcelable("userIn");
                    String userWarehouseNo = userBundle.getString("warehouseNo");
                    mOnGetWarningDataFinish.onUserChange(userIn, userWarehouseNo, isOpenDoor);
                    break;
                case USER_OUT:
                    Bundle userOutBundle = msg.getData();
                    UserOutVM userOut = userOutBundle.getParcelable("userOut");
                    String userOutWarehouseNo = userOutBundle.getString("warehouseNo");
                    if (null == userOut) {
                        break;
                    }
                    mOnGetWarningDataFinish.onUserOut(userOutWarehouseNo, userOut);
                    break;
                case ENTRANCE_GUARD_STATE:
                    Bundle doorState = msg.getData();
                    String doorWarehouseNo = doorState.getString("warehouseNo");
                    EntranceGuardState entranceGuardState = doorState.getParcelable("entranceGuardState");
                    mOnGetWarningDataFinish.onReceiveDoorState(doorWarehouseNo, entranceGuardState);
                    break;
                case SHELF_STATE:
                    Bundle shelfStateBundle = msg.getData();
                    String shelfWarehouseNo = shelfStateBundle.getString("warehouseNo");
                    ShelfState shelfState = shelfStateBundle.getParcelable("shelfState");
                    mOnGetWarningDataFinish.onReceiveShelfState(shelfWarehouseNo, shelfState);
                    break;
                default:
                    break;
            }
        }
    }
}

File: AndroidStudioProjects/app/src/main/java/cn/zemic/hy/display/unmannedstoragedisplay/presenter/imp/alarm/AlarmRepository.kt
package cn.zemic.hy.display.unmannedstoragedisplay.presenter.imp.alarm

import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.util.Log
import cn.zemic.hy.display.unmannedstoragedisplay.model.bean.FilterLedForMobileUiCommand
import cn.zemic.hy.display.unmannedstoragedisplay.model.viewmodel.*
import cn.zemic.hy.display.unmannedstoragedisplay.network.URLFactory
import cn.zemic.hy.display.unmannedstoragedisplay.network.service.FetchDataService
import cn.zemic.hy.display.unmannedstoragedisplay.presenter.interfaces.IAlarmRepository
import cn.zemic.hy.display.unmannedstoragedisplay.presenter.interfaces.IBaseRepository
import microsoft.aspnet.signalr.client.Platform
import microsoft.aspnet.signalr.client.http.android.AndroidPlatformComponent
import microsoft.aspnet.signalr.client.hubs.HubConnection
import microsoft.aspnet.signalr.client.hubs.HubProxy
import microsoft.aspnet.signalr.client.transport.ServerSentEventsTransport
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*
import java.util.concurrent.ExecutionException

/**
 * @author fxs
 */
class AlarmRepository : IAlarmRepository {
    private var mHandler: MyMessagesHandler? = null
    private var mHubProxy: HubProxy? = null

    override fun getWareHouseNo(deviceId: String, onGetWareHouseNoFinish: IBaseRepository.OnGetWareHouseNoFinish) {
        val filter = FilterLedForMobileUiCommand(deviceId)
        FetchDataService.getWareHouseNoService(filter).enqueue(object : Callback<String> {
            override fun onResponse(call: Call<String>, response: Response<String>) {
                if (!response.isSuccessful) {
                    onGetWareHouseNoFinish.onFail(String.format(Locale.CHINA, "获取仓库号响应异常，错误码：%d", response.code()))
                    return
                }
                val wareHouseNo = response.body()
                onGetWareHouseNoFinish.onSuccess(wareHouseNo)
            }

            override fun onFailure(call: Call<String>, t: Throwable) {
                onGetWareHouseNoFinish.onFail(if (t.message == null || "" == t.message) "获取仓库号失败" else t.message)
            }
        })
    }

    override fun connectServiceForGetWarningAndUserInfo(deviceId: String, wareHouseNo: String, onGetWarningDataFinish: IBaseRepository.OnGetWarningDataFinish) {
        mOnGetWarningDataFinish = onGetWarningDataFinish
        mHandler = MyMessagesHandler()
        Platform.loadPlatformComponent(AndroidPlatformComponent())
        val serverURI = String.format(Locale.CHINA, "%ssignalr", URLFactory.serverURI)
        //        String serverURI = "http://192.168.0.2:25535/signalr";
        val connection = object : HubConnection(serverURI) {
            override fun onConnected() {
                super.onConnected()
                Log.i("hubOnConnected", "hubOnConnected")
            }

            override fun onError(error: Throwable, mustCleanCurrentConnection: Boolean) {
                super.onError(error, mustCleanCurrentConnection)
                val message = Message.obtain()
                message.what = DATA_ERROR
                message.obj = if ("" == error.message) "连接推送服务器出现异常，请联系管理员" else error.message
                mHandler!!.sendMessage(message)
            }

            override fun onClosed() {
                super.onClosed()
            }
        }
        //hub agency
        mHubProxy = connection.createHubProxy("electronicSignsHub")
        //hubProxy.on可以理解为讯息or事件监听器
        mHubProxy!!.run {
            on("sendUserIn", this@AlarmRepository::runWithUserInfo, String::class.java, Boolean::class.java, UserInViewModel::class.java, MachineBindDoorControlViewModel::class.java)
            on("sendCheck", this@AlarmRepository::runWithIsCheck, String::class.java, Boolean::class.java)
            on("sendMaintain", this@AlarmRepository::runWithMaintain, String::class.java, Boolean::class.java)
            on("sendTemAndHum", this@AlarmRepository::runWithTemperature, String::class.java, Float::class.java, Float::class.java)
            on("sendUserLeave", this@AlarmRepository::runWithUserOut, String::class.java, UserOutVM::class.java)
            //接收告警信息
            on("sendWarnInform", this@AlarmRepository::runWithWarnInfo, String::class.java, OperateWarningInformViewModel::class.java)
            //接受日志信息
            on("sendConnectMessage", this@AlarmRepository::runWithConnectMsg, String::class.java, String::class.java, Boolean::class.java)
            on("sendAttendanceMachine", this@AlarmRepository::entranceGuardState, String::class.java, EntranceGuardState::class.java)
            on("sendLedShelfAndunit", this@AlarmRepository::shelfState, String::class.java, ShelfState::class.java)
        }
        val con = connection.start(ServerSentEventsTransport(connection.logger))
        try {
            con.get()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        }

        con.onError { error -> onError(error) }
        if (con.isDone) {
            //调用服务器连接方法
            mHubProxy!!.invoke("Connect", wareHouseNo, deviceId)
        }
    }

    override fun getHubProxy(): HubProxy {
        if (null == mHubProxy) {
            throw NullPointerException("mHubProxy is null")
        }
        return mHubProxy!!
    }

    private fun runWithUserInfo(warehouseNo: String, isOpenDoor: Boolean?, userIn: UserInViewModel, doorControl: MachineBindDoorControlViewModel) {
        val message = Message.obtain()
        message.what = USER_CHANGED
        val bundle = Bundle()
        bundle.putBoolean("isOpenDoor", isOpenDoor!!)
        bundle.putString("warehouseNo", warehouseNo)
        bundle.putParcelable("userIn", userIn)
        bundle.putParcelable("doorControl", doorControl)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }


    private fun runWithUserOut(warehouseNo: String, user: UserOutVM) {
        val message = Message.obtain()
        message.what = USER_OUT
        val bundle = Bundle()
        bundle.putString("warehouseNo", warehouseNo)
        bundle.putParcelable("userOut", user)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun runWithIsCheck(warehouseNo: String, isCheck: Boolean?) {
        val message = Message.obtain()
        message.what = CHECK
        val bundle = Bundle()
        bundle.putBoolean("isCheck", isCheck!!)
        bundle.putString("warehouseNo", warehouseNo)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun runWithMaintain(warehouseNo: String, isMaintain: Boolean?) {
        val message = Message.obtain()
        message.what = MAINTAIN
        val bundle = Bundle()
        bundle.putBoolean("isMaintain", isMaintain!!)
        bundle.putString("warehouseNo", warehouseNo)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun runWithTemperature(wareHouseNo: String, temperature: Float, humidity: Float) {
        val message = Message.obtain()
        message.what = TEMPERATURE_HUMIDITY
        val bundle = Bundle()
        bundle.putString("wareHouseNo", wareHouseNo)
        bundle.putFloat("temperature", temperature)
        bundle.putFloat("humidity", humidity)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun runWithWarnInfo(warehouseNo: String, warningInform: OperateWarningInformViewModel) {
        val message = Message.obtain()
        message.what = DATA_CHANGED
        val bundle = Bundle()
        bundle.putString("warehouseNo", warehouseNo)
        bundle.putParcelable("warningInform", warningInform)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun runWithConnectMsg(connectionMsg: String, connectionId: String, isConnect: Boolean) {
        val message = Message.obtain()
        message.what = DATA_SEND_MSG
        val bundle = Bundle()
        if (isConnect) {
            mConnectionId = connectionId
        }
        bundle.putString("connectionId", connectionId)
        bundle.putBoolean("isConnect", isConnect)
        message.data = bundle
        message.obj = connectionMsg
        mHandler!!.sendMessage(message)
    }

    private fun entranceGuardState(warehouseNo: String, entranceGuardState: EntranceGuardState) {
        val message = Message.obtain()
        message.what = ENTRANCE_GUARD_STATE
        val bundle = Bundle()
        bundle.putString("warehouseNo", warehouseNo)
        bundle.putParcelable("entranceGuardState", entranceGuardState)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun shelfState(warehouseNo: String, shelfState: ShelfState) {
        val message = Message.obtain()
        message.what = SHELF_STATE
        val bundle = Bundle()
        bundle.putString("warehouseNo", warehouseNo)
        bundle.putParcelable("shelfState", shelfState)
        message.data = bundle
        mHandler!!.sendMessage(message)
    }

    private fun onError(error: Throwable) {
        val message = Message.obtain()
        message.what = DATA_ERROR
        message.obj = if ("" == error.message) error.message else "连接中断，请联系管理员"
        mHandler!!.sendMessage(message)
    }

    private class MyMessagesHandler : Handler() {
        private val LIST_MAX_SIZE = 10
        private val mOperateWarningInformViewModels = ArrayList<OperateWarningInformViewModel>()

        override fun handleMessage(msg: Message) {
            if (mOnGetWarningDataFinish == null) {
                return
            }
            super.handleMessage(msg)
            when (msg.what) {
                TEMPERATURE_HUMIDITY -> {
                    val bundletempAndHum = msg.data
                    val temperature = bundletempAndHum.getFloat("temperature")
                    val humidity = bundletempAndHum.getFloat("humidity")
                    val wareHouseForTemp = bundletempAndHum.getString("wareHouseNo")
                    mOnGetWarningDataFinish!!.onFetchTemperatureAndHumidityFinish(wareHouseForTemp, temperature, humidity)
                }
                MAINTAIN -> {
                    val bundleMaintain = msg.data
                    val warehouseNoForMaintain = bundleMaintain.getString("warehouseNo")
                    val isMaintain = bundleMaintain.getBoolean("isMaintain")
                    mOnGetWarningDataFinish!!.onRunningMaintain(warehouseNoForMaintain, isMaintain)
                }
                CHECK -> {
                    val bundleCheck = msg.data
                    val warehouseNo = bundleCheck.getString("warehouseNo")
                    val isCheck = bundleCheck.getBoolean("isCheck")
                    mOnGetWarningDataFinish!!.onRunningCheck(warehouseNo, isCheck)
                }

                DATA_SEND_MSG -> {
                    val bundleMsg = msg.data
                    val connectionId = bundleMsg.getString("connectionId")
                    val isConnect = bundleMsg.getBoolean("isConnect")
                    Log.e("DATA_SEND_MSG", msg.obj.toString() + connectionId!!)
                    if (isConnect) {
                        mOnGetWarningDataFinish!!.onShow(msg.obj.toString())
                    } else {
                        if (mConnectionId == connectionId) {
                            mOnGetWarningDataFinish!!.onShow(msg.obj.toString())
                        }
                    }
                }
                DATA_ERROR -> {
                    Log.e("DATA_ERROR", msg.obj.toString())
                    mOnGetWarningDataFinish!!.onShow(msg.obj.toString())
                }
                DATA_CHANGED -> {
                    val bundle = msg.data
                    val wainingInfo = bundle.getParcelable<OperateWarningInformViewModel>("warningInform")
                    if (wainingInfo != null) {
                        if (mOperateWarningInformViewModels.size < LIST_MAX_SIZE) {
                            mOperateWarningInformViewModels.add(0, wainingInfo)
                        } else {
                            mOperateWarningInformViewModels.removeAt(mOperateWarningInformViewModels.size - 1)
                            mOperateWarningInformViewModels.add(0, wainingInfo)
                        }
                    }
                    //update view
                    if (mOperateWarningInformViewModels.size <= 1) {
                        mOnGetWarningDataFinish!!.onSuccess(mOperateWarningInformViewModels, true)
                    } else {
                        mOnGetWarningDataFinish!!.onSuccess(mOperateWarningInformViewModels, false)
                    }
                }
                USER_CHANGED -> {
                    val userBundle = msg.data
                    val isOpenDoor = userBundle.getBoolean("isOpenDoor")
                    val userIn = userBundle.getParcelable<UserInViewModel>("userIn")
                    val userWarehouseNo = userBundle.getString("warehouseNo")
                    mOnGetWarningDataFinish!!.onUserChange(userIn, userWarehouseNo, isOpenDoor)
                }
                USER_OUT -> {
                    val userOutBundle = msg.data
                    val userOut = userOutBundle.getParcelable<UserOutVM>("userOut")
                    val userOutWarehouseNo = userOutBundle.getString("warehouseNo")
                    if (null == userOut) {
                    }
                    mOnGetWarningDataFinish!!.onUserOut(userOutWarehouseNo, userOut)
                }
                ENTRANCE_GUARD_STATE -> {
                    val doorState = msg.data
                    val doorWarehouseNo = doorState.getString("warehouseNo")
                    val entranceGuardState = doorState.getParcelable<EntranceGuardState>("entranceGuardState")
                    mOnGetWarningDataFinish!!.onReceiveDoorState(doorWarehouseNo, entranceGuardState)
                }
                SHELF_STATE -> {
                    val shelfStateBundle = msg.data
                    var shelfWarehouseNo = shelfStateBundle.getString("warehouseNo")
                    shelfWarehouseNo = shelfWarehouseNo ?: ""
                    val shelfState = shelfStateBundle.getParcelable<ShelfState>("shelfState")
                    shelfState?.Shelfs?.sortBy {
                        it.shelfNo
                    }
                    mOnGetWarningDataFinish!!.onReceiveShelfState(shelfWarehouseNo, shelfState)
                }
                else -> {
                }
            }
        }
    }

    companion object {
        private const val DATA_ERROR = 0
        private const val DATA_CHANGED = 1
        private const val USER_CHANGED = 2
        private const val DATA_SEND_MSG = 3
        private const val USER_OUT = 4
        private const val CHECK = 5
        private const val MAINTAIN = 6
        private const val TEMPERATURE_HUMIDITY = 7
        private const val ENTRANCE_GUARD_STATE = 8
        private const val SHELF_STATE = 9

        private var mOnGetWarningDataFinish: IBaseRepository.OnGetWarningDataFinish? = null
        private var mConnectionId: String? = null
    }
}

